tests/zero-downtime-scaling/zero-downtime-scaling.md

```diff
@@ -0,0 +1,170 @@
+    (however small) periods of downtime.
+  - *curl* will generate 1 request every 0.1s. While it might not catch small periods of downtime, it will
+    give us timeline of failed request for big periods of downtime, which wrk doesn't do.
+
+## Tests
+
+Run the following tests on both the 10 and 25 node clusters.
+
+### Setup
+
+- Create a cluster.
+- Deploy NGF, the latest stable version with 1 replicas.
+  - If testing with the 25 node cluster, add a pod affinity rule.
+- Expose NGF via a Service Load Balancer, internal (only accessible within the Google Cloud region) by
+  adding `networking.gke.io/load-balancer-type: "Internal"` annotation to the Service.
+- Deploy backend apps.
+- Configure Gateway.
+- Expose apps via HTTPRoutes.
+- Check statuses of the Gateway and HTTPRoutes.
+- Start sending traffic using wrk from tester VMs: HTTP and HTTPs.
+- Check that NGINX access logs look good (all responses 200).
+- Check that there are no errors in NGINX errors logs and NGF error logs.
+
+### Scale Gradually
+
+Do the following until there are 25 NGF Pods running:
+
+- Scale up by 1 Pod.
+- Wait for the new Pod to become ready.
+- Update the Gateway resource - add one listener - and make sure NGF processed the update (it updates the status of the
+  Gateway resource accordingly).
+
+The, do the following until there are 1 NGF Pods running:
+
+- Scale down by 1 Pod.
+- Wait for the old Pod to terminate.
+- Update the Gateway resource - remove one listener - and make sure NGF processed the update (it updates the status of the
+  Gateway resource accordingly).
+
+Now, [analyze](#analyze) the results.
+
+### Scale Abruptly
+
+Scale up:
+
+- Scale up to 25 Pods.
+- Wait for the new Pods to become ready.
+- Update the Gateway resource - add one listener - and make sure NGF processed the update (it updates the status of the
+   Gateway resource accordingly).
+
+Scale down:
+
+- Scale up to 25 Pods.
+- Wait for the new Pods to become ready.
+- Update the Gateway resource - remove one listener - and make sure NGF processed the update (it updates the status of the
+  Gateway resource accordingly).
+
+Now, [analyze](#analyze) the results.
+
+### Analyze
+
+After each test analyze and record the following:
+
+- Tester VMs:
+  - Analyze the output of wrk commands for errors and latencies.
+  - Create graphs from curl output (TODO: add instructions).
+- If scaling down, check the old Pods logs in Google Monitoring:
+  - NGINX Access logs - we expect only 200 responses.
+    Google Monitoring query:
+
+      ```text
+      severity=INFO
+      "GET" "HTTP/1.1" -"200"
+      ```
+
+  - NGINX Error logs - we expect no errors or warnings:
+    Google Monitoring query:
+
+      ```text
+      severity=ERROR
+      SEARCH("`[warn]`") OR SEARCH("`[error]`")
+      ```
+
+  - NGF logs - we expect no errors.
+  - Specifically look at the NGF logs before it exited, to make sure all components shutdown correctly.
```

Comment:
since there will be 25 pods, maybe there is a way to analyze NGF logs using queries only rather than inspecting each pod logs individually

Reply:
Good idea. I'm sure there's a way to do that with GKE